src/harness/util.ts

```diff
@@ -101,6 +101,7 @@ export function addUTF8ByteOrderMark(text: string) {
     return getByteOrderMarkLength(text) === 0 ? "\u00EF\u00BB\u00BF" + text : text;
 }
 
+/** @knipignore -- knip doesn't understand this in src/testRunner/unittests/semver.ts: import theory = Utils.theory; */

```

Comment:
Interesting; what kind of analysis is failing? I thought it was just using TS under the hood?

Reply:
Prior to Knip v4, it relied heavily on `findReferences` indeed. But that got too slow for large repos (esp. monorepos). Since 99% of exports consumption is basically either direct or property access, Knip gets away with simpler heuristics. I remember this brief conversation regarding basically the exact same topic: https://discord.com/channels/508357248330760243/640177429775777792/1196882108870823978 (still planning to give this a shot: "findReferences of imports within same file")

Reply:
Btw for completeness sake, memory consumption in Knip v3 was also high (many files in the program/memory):

- https://knip.dev/blog/slim-down-to-speed-up
- https://knip.dev/blog/knip-v4